push_notifications/webpush.py

```diff
@@ -1,39 +1,55 @@
+import warnings
+
 from pywebpush import WebPushException, webpush
 
 from .conf import get_manager
 from .exceptions import WebPushError
 
 
 def get_subscription_info(application_id, uri, browser, auth, p256dh):
-	url = get_manager().get_wp_post_url(application_id, browser)
+	if uri.startswith("https://"):
+		endpoint = uri
+	else:
+		url = get_manager().get_wp_post_url(application_id, browser)
+		endpoint = "{}/{}".format(url, uri)
+		warnings.warn(
+			"registratipn_id should be the full endpoint returned from pushManager.subscribe",
+			DeprecationWarning,
+			stacklevel=2,
+		)
 	return {
-		"endpoint": "{}/{}".format(url, uri),
+		"endpoint": endpoint,
 		"keys": {
 			"auth": auth,
 			"p256dh": p256dh,
 		}
 	}
 
 
-def webpush_send_message(
-	uri, message, browser, auth, p256dh, application_id=None, **kwargs
-):
-	subscription_info = get_subscription_info(application_id, uri, browser, auth, p256dh)
-
+def webpush_send_message(device, message, **kwargs):
+	subscription_info = get_subscription_info(
+		device.application_id, device.registration_id,
+		device.browser, device.auth, device.p256dh)
 	try:
 		response = webpush(
 			subscription_info=subscription_info,
 			data=message,
-			vapid_private_key=get_manager().get_wp_private_key(application_id),
-			vapid_claims=get_manager().get_wp_claims(application_id).copy(),
+			vapid_private_key=get_manager().get_wp_private_key(device.application_id),
+			vapid_claims=get_manager().get_wp_claims(device.application_id).copy(),
 			**kwargs
 		)
-		results = {"results": [{}]}
-		if not response.ok:
-			results["results"][0]["error"] = response.content
-			results["results"][0]["original_registration_id"] = response.content
-		else:
+		results = {"results": [{"original_registration_id": device.registration_id}]}
+		if response.ok:
 			results["success"] = 1
+		else:
+			results["failure"] = 1
+			results["results"][0]["error"] = response.content
 		return results
 	except WebPushException as e:
+		if e.response is not None and e.response.status_code in [404, 410]:
+			results["failure"] = 1
```

Comment:
I get an error here because the exception comes from line 34, but `results` is defined on line 41. It should be enough to just switch them around.

Reply:
Excellent point, commit added to address it.

Reply:
Could you add a comment explaining the two status codes and why you are disabling the device

Reply:
how about?
```suggestion
			# if we find that the status from the vendor was 404 (invalide url) or 410 (Gone) we should no
			# longer try to send messages to this device.
			# https://web.dev/push-notifications-common-issues-and-reporting-bugs/#http-status-codes
			results["failure"] = 1
```